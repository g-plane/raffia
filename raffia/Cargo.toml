[package]
name = "raffia"
version = "0.1.0"
edition = "2021"
authors = ["Pig Fang <g-plane@hotmail.com>"]
description = "One parser for CSS/SCSS/Sass/Less."
repository = "https://github.com/g-plane/raffia"
license = "MIT"

[dependencies]
raffia_macro = { path = "../raffia_macro", version = "0.1" }
serde = { version = "1.0", features = ["derive"], optional = true }
smallvec = { version = "1.9", features = ["const_generics", "serde"] }

[dev-dependencies]
codespan-reporting = "0.11"
criterion = "0.3"
insta = { version = "1.19", features = ["glob", "ron"] }
lightningcss = "1.0.0-alpha.35"
serde_json = "1.0"
swc_common = "0.29.0"
swc_css_ast = "0.114.0"
swc_css_parser = "0.123.0"

[features]
serialize = ["serde"]

[[bench]]
name = "parse_selector"
harness = false

[[bench]]
name = "parse_stylesheet"
harness = false
