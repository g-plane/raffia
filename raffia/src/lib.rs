//! Raffia is a parser can parse CSS, SCSS, Sass (indented syntax) and Less.
//!
//! ## Basic Usage
//!
//! This crate provides a simple API to get started.
//!
//! First, create a parser, give it the source code and specify the syntax,
//! then call the [`parse`](Parser::parse) method:
//!
//! ```rust
//! use raffia::{ast::Stylesheet, Parser, Syntax};
//!
//! let mut parser = Parser::new("a {}", Syntax::Css); // syntax can also be `Scss`, `Sass` or `Less`
//! let result = parser.parse::<Stylesheet>();
//! match result {
//!     Ok(ast) => {
//!         // parsed successfully
//!         println!("{:#?}", ast);
//!     }
//!     Err(error) => {
//!         // it failed, error message and position can be accessed via `error`
//!         println!("{:#?}", error);
//!     }
//! }
//! ```
//!
//! ## Advanced Usage
//!
//! ### More Parser Options
//!
//! If you need to control more parser options, you can use [`ParserBuilder`].
//!
//! For example, to collect comments:
//!
//! ```rust
//! use raffia::ParserBuilder;
//!
//! let mut comments = vec![];
//! let builder = ParserBuilder::new("/* comment */ a {}").comments(&mut comments);
//! let mut parser = builder.build();
//! ```
//!
//! By default, syntax is CSS when using parser builder. You can customize it:
//!
//! ```rust
//! use raffia::{ParserBuilder, Syntax};
//!
//! let builder = ParserBuilder::new("a {}").syntax(Syntax::Scss);
//! ```
//!
//! ### Parse Partial Structure
//!
//! Sometimes you don't want to parse a full stylesheet.
//! Say you only need to parse a qualified rule or even a single declaration.
//! All you need to do is to update the generics of the [`parse`](Parser::parse) method.
//!
//! ```rust
//! use raffia::{ast::QualifiedRule, Parser, Syntax};
//!
//! let mut parser = Parser::new("a {}", Syntax::Css);
//! parser.parse::<QualifiedRule>();
//! ```
//!
//! and
//!
//! ```rust
//! use raffia::{ast::Declaration, Parser, Syntax};
//!
//! let mut parser = Parser::new("color: green", Syntax::Css);
//! parser.parse::<Declaration>();
//! ```
//!
//! Not all AST nodes support the usage above;
//! technically, those nodes that implement [`Parse`] trait are supported.
//!
//! ### Retrieve Recoverable Errors
//!
//! There may be some recoverable errors which doesn't affect on producing AST.
//! To retrieve those errors, use [`recoverable_errors`](Parser::recoverable_errors).
//!
//! ```rust
//! use raffia::{ast::Stylesheet, Parser, Syntax};
//!
//! let mut parser = Parser::new("@keyframes kf { invalid {} }", Syntax::Css);
//! let result = parser.parse::<Stylesheet>();
//! assert!(result.is_ok());
//! println!("{:?}", parser.recoverable_errors());
//! ```
//!
//! ## Serialization
//!
//! Produced AST can be serialized by Serde, but this feature is disabled by default.
//! You need to enable feature `serialize` manually:
//!
//! ```toml
//! raffia = { version = "*", features = ["serialize"] }
//! ```
//!
//! Then you can pass AST to Serde.
//!
//! Note that Raffia only supports serialization. Deserialization isn't supported.

pub use config::{ParserOptions, Syntax};
pub use parser::{Parse, Parser, ParserBuilder};
pub use pos::{Span, Spanned};
pub use span_ignored_eq::SpanIgnoredEq;
pub use tokenizer::token;

pub mod ast;
mod config;
pub mod error;
mod parser;
pub mod pos;
mod span_ignored_eq;
mod tokenizer;
mod util;
