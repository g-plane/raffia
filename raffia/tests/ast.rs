use codespan_reporting::{
    diagnostic::{Diagnostic, Label},
    files::SimpleFile,
    term::{self, termcolor::Buffer},
};
use insta::{assert_ron_snapshot, glob, Settings};
use raffia::{ast::Stylesheet, Parser, Syntax};
use std::fs;

#[test]
fn ast_snapshot() {
    glob!("ast/**/*.{css,scss,sass,less}", |path| {
        let code = fs::read_to_string(path).unwrap();
        let syntax = match path.extension().unwrap().to_str().unwrap() {
            "css" => Syntax::Css,
            "scss" => Syntax::Scss,
            "sass" => Syntax::Sass,
            "less" => Syntax::Less,
            _ => unreachable!("unknown file extension"),
        };
        let mut parser = Parser::new(&code, syntax);
        let ast = match parser.parse::<Stylesheet>() {
            Ok(ast) => {
                let recoverable_errors = parser.recoverable_errors();
                assert!(
                    recoverable_errors.is_empty(),
                    "'{}' has recoverable errors: {recoverable_errors:?}",
                    path.file_name().unwrap().to_str().unwrap(),
                );
                ast
            }
            Err(error) => {
                let file = SimpleFile::new(path.file_name().unwrap().to_str().unwrap(), &code);
                let diagnostic = Diagnostic::error()
                    .with_message(error.kind.to_string())
                    .with_labels(vec![Label::primary((), error.span.start..error.span.end)]);
                let mut buffer = Buffer::ansi();
                let config = term::Config::default();
                term::emit(&mut buffer, &config, &file, &diagnostic).unwrap();
                panic!("\n{}", String::from_utf8(buffer.into_inner()).unwrap());
            }
        };

        let mut settings = Settings::clone_current();
        settings.set_snapshot_path(path.parent().unwrap());
        settings.remove_snapshot_suffix();
        settings.set_prepend_module_to_snapshot(false);
        settings.remove_input_file();
        settings.set_omit_expression(true);
        settings.remove_input_file();
        settings.remove_info();
        settings.bind(|| {
            let name = path.file_stem().unwrap().to_str().unwrap();
            assert_ron_snapshot!(name, ast);
        });
    });
}
