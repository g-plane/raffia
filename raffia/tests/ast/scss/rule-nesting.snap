---
source: raffia/tests/ast.rs
assertion_line: 54
snapshot_kind: text
---
Stylesheet(
  type: "Stylesheet",
  statements: [
    QualifiedRule(
      type: "QualifiedRule",
      selector: SelectorList(
        type: "SelectorList",
        selectors: [
          ComplexSelector(
            type: "ComplexSelector",
            children: [
              CompoundSelector(
                type: "CompoundSelector",
                children: [
                  TagNameSelector(
                    type: "TagNameSelector",
                    name: WqName(
                      type: "WqName",
                      name: Ident(
                        type: "Ident",
                        name: "a",
                        raw: "a",
                        span: Span(
                          start: 0,
                          end: 1,
                        ),
                      ),
                      prefix: None,
                      span: Span(
                        start: 0,
                        end: 1,
                      ),
                    ),
                    span: Span(
                      start: 0,
                      end: 1,
                    ),
                  ),
                ],
                span: Span(
                  start: 0,
                  end: 1,
                ),
              ),
            ],
            span: Span(
              start: 0,
              end: 1,
            ),
          ),
        ],
        commaSpans: [],
        span: Span(
          start: 0,
          end: 1,
        ),
      ),
      block: SimpleBlock(
        type: "SimpleBlock",
        statements: [
          Declaration(
            type: "Declaration",
            name: Ident(
              type: "Ident",
              name: "color",
              raw: "color",
              span: Span(
                start: 8,
                end: 13,
              ),
            ),
            colonSpan: Span(
              start: 13,
              end: 14,
            ),
            value: [
              Ident(
                type: "Ident",
                name: "red",
                raw: "red",
                span: Span(
                  start: 15,
                  end: 18,
                ),
              ),
            ],
            important: None,
            lessPropertyMerge: None,
            span: Span(
              start: 8,
              end: 18,
            ),
          ),
          QualifiedRule(
            type: "QualifiedRule",
            selector: SelectorList(
              type: "SelectorList",
              selectors: [
                ComplexSelector(
                  type: "ComplexSelector",
                  children: [
                    CompoundSelector(
                      type: "CompoundSelector",
                      children: [
                        TagNameSelector(
                          type: "TagNameSelector",
                          name: WqName(
                            type: "WqName",
                            name: Ident(
                              type: "Ident",
                              name: "b",
                              raw: "b",
                              span: Span(
                                start: 24,
                                end: 25,
                              ),
                            ),
                            prefix: None,
                            span: Span(
                              start: 24,
                              end: 25,
                            ),
                          ),
                          span: Span(
                            start: 24,
                            end: 25,
                          ),
                        ),
                      ],
                      span: Span(
                        start: 24,
                        end: 25,
                      ),
                    ),
                  ],
                  span: Span(
                    start: 24,
                    end: 25,
                  ),
                ),
              ],
              commaSpans: [],
              span: Span(
                start: 24,
                end: 25,
              ),
            ),
            block: SimpleBlock(
              type: "SimpleBlock",
              statements: [
                QualifiedRule(
                  type: "QualifiedRule",
                  selector: SelectorList(
                    type: "SelectorList",
                    selectors: [
                      ComplexSelector(
                        type: "ComplexSelector",
                        children: [
                          CompoundSelector(
                            type: "CompoundSelector",
                            children: [
                              TagNameSelector(
                                type: "TagNameSelector",
                                name: WqName(
                                  type: "WqName",
                                  name: Ident(
                                    type: "Ident",
                                    name: "c",
                                    raw: "c",
                                    span: Span(
                                      start: 36,
                                      end: 37,
                                    ),
                                  ),
                                  prefix: None,
                                  span: Span(
                                    start: 36,
                                    end: 37,
                                  ),
                                ),
                                span: Span(
                                  start: 36,
                                  end: 37,
                                ),
                              ),
                            ],
                            span: Span(
                              start: 36,
                              end: 37,
                            ),
                          ),
                        ],
                        span: Span(
                          start: 36,
                          end: 37,
                        ),
                      ),
                    ],
                    commaSpans: [],
                    span: Span(
                      start: 36,
                      end: 37,
                    ),
                  ),
                  block: SimpleBlock(
                    type: "SimpleBlock",
                    statements: [
                      QualifiedRule(
                        type: "QualifiedRule",
                        selector: SelectorList(
                          type: "SelectorList",
                          selectors: [
                            ComplexSelector(
                              type: "ComplexSelector",
                              children: [
                                CompoundSelector(
                                  type: "CompoundSelector",
                                  children: [
                                    TagNameSelector(
                                      type: "TagNameSelector",
                                      name: WqName(
                                        type: "WqName",
                                        name: Ident(
                                          type: "Ident",
                                          name: "d",
                                          raw: "d",
                                          span: Span(
                                            start: 52,
                                            end: 53,
                                          ),
                                        ),
                                        prefix: None,
                                        span: Span(
                                          start: 52,
                                          end: 53,
                                        ),
                                      ),
                                      span: Span(
                                        start: 52,
                                        end: 53,
                                      ),
                                    ),
                                  ],
                                  span: Span(
                                    start: 52,
                                    end: 53,
                                  ),
                                ),
                              ],
                              span: Span(
                                start: 52,
                                end: 53,
                              ),
                            ),
                          ],
                          commaSpans: [],
                          span: Span(
                            start: 52,
                            end: 53,
                          ),
                        ),
                        block: SimpleBlock(
                          type: "SimpleBlock",
                          statements: [
                            Declaration(
                              type: "Declaration",
                              name: Ident(
                                type: "Ident",
                                name: "color",
                                raw: "color",
                                span: Span(
                                  start: 72,
                                  end: 77,
                                ),
                              ),
                              colonSpan: Span(
                                start: 77,
                                end: 78,
                              ),
                              value: [
                                Ident(
                                  type: "Ident",
                                  name: "red",
                                  raw: "red",
                                  span: Span(
                                    start: 79,
                                    end: 82,
                                  ),
                                ),
                              ],
                              important: None,
                              lessPropertyMerge: None,
                              span: Span(
                                start: 72,
                                end: 82,
                              ),
                            ),
                          ],
                          span: Span(
                            start: 54,
                            end: 97,
                          ),
                        ),
                        span: Span(
                          start: 52,
                          end: 97,
                        ),
                      ),
                    ],
                    span: Span(
                      start: 38,
                      end: 107,
                    ),
                  ),
                  span: Span(
                    start: 36,
                    end: 107,
                  ),
                ),
                Declaration(
                  type: "Declaration",
                  name: Ident(
                    type: "Ident",
                    name: "color",
                    raw: "color",
                    span: Span(
                      start: 116,
                      end: 121,
                    ),
                  ),
                  colonSpan: Span(
                    start: 121,
                    end: 122,
                  ),
                  value: [
                    Ident(
                      type: "Ident",
                      name: "red",
                      raw: "red",
                      span: Span(
                        start: 123,
                        end: 126,
                      ),
                    ),
                  ],
                  important: None,
                  lessPropertyMerge: None,
                  span: Span(
                    start: 116,
                    end: 126,
                  ),
                ),
              ],
              span: Span(
                start: 26,
                end: 133,
              ),
            ),
            span: Span(
              start: 24,
              end: 133,
            ),
          ),
          QualifiedRule(
            type: "QualifiedRule",
            selector: SelectorList(
              type: "SelectorList",
              selectors: [
                ComplexSelector(
                  type: "ComplexSelector",
                  children: [
                    CompoundSelector(
                      type: "CompoundSelector",
                      children: [
                        TagNameSelector(
                          type: "TagNameSelector",
                          name: WqName(
                            type: "WqName",
                            name: Ident(
                              type: "Ident",
                              name: "e",
                              raw: "e",
                              span: Span(
                                start: 138,
                                end: 139,
                              ),
                            ),
                            prefix: None,
                            span: Span(
                              start: 138,
                              end: 139,
                            ),
                          ),
                          span: Span(
                            start: 138,
                            end: 139,
                          ),
                        ),
                      ],
                      span: Span(
                        start: 138,
                        end: 139,
                      ),
                    ),
                  ],
                  span: Span(
                    start: 138,
                    end: 139,
                  ),
                ),
              ],
              commaSpans: [],
              span: Span(
                start: 138,
                end: 139,
              ),
            ),
            block: SimpleBlock(
              type: "SimpleBlock",
              statements: [
                Declaration(
                  type: "Declaration",
                  name: Ident(
                    type: "Ident",
                    name: "color",
                    raw: "color",
                    span: Span(
                      start: 150,
                      end: 155,
                    ),
                  ),
                  colonSpan: Span(
                    start: 155,
                    end: 156,
                  ),
                  value: [
                    Ident(
                      type: "Ident",
                      name: "red",
                      raw: "red",
                      span: Span(
                        start: 157,
                        end: 160,
                      ),
                    ),
                  ],
                  important: None,
                  lessPropertyMerge: None,
                  span: Span(
                    start: 150,
                    end: 160,
                  ),
                ),
              ],
              span: Span(
                start: 140,
                end: 167,
              ),
            ),
            span: Span(
              start: 138,
              end: 167,
            ),
          ),
          QualifiedRule(
            type: "QualifiedRule",
            selector: SelectorList(
              type: "SelectorList",
              selectors: [
                ComplexSelector(
                  type: "ComplexSelector",
                  children: [
                    CompoundSelector(
                      type: "CompoundSelector",
                      children: [
                        TagNameSelector(
                          type: "TagNameSelector",
                          name: WqName(
                            type: "WqName",
                            name: Ident(
                              type: "Ident",
                              name: "foo",
                              raw: "foo",
                              span: Span(
                                start: 172,
                                end: 175,
                              ),
                            ),
                            prefix: None,
                            span: Span(
                              start: 172,
                              end: 175,
                            ),
                          ),
                          span: Span(
                            start: 172,
                            end: 175,
                          ),
                        ),
                      ],
                      span: Span(
                        start: 172,
                        end: 175,
                      ),
                    ),
                  ],
                  span: Span(
                    start: 172,
                    end: 175,
                  ),
                ),
              ],
              commaSpans: [],
              span: Span(
                start: 172,
                end: 175,
              ),
            ),
            block: SimpleBlock(
              type: "SimpleBlock",
              statements: [],
              span: Span(
                start: 176,
                end: 178,
              ),
            ),
            span: Span(
              start: 172,
              end: 178,
            ),
          ),
        ],
        span: Span(
          start: 2,
          end: 180,
        ),
      ),
      span: Span(
        start: 0,
        end: 180,
      ),
    ),
  ],
  span: Span(
    start: 0,
    end: 181,
  ),
)
